### PR TITLE
[oneDNN] Keras LayerNormalization fusion with oneDNN CPU backend

### PR DESCRIPTION
**Pull request by mdfaijul**
Keras LayerNormalization API creates a set op smaller ops that can be realized by a single operation with oneDNN library on CPU. This PR fuses smaller ops into a single op using grappler remapper optimizer. Current fusion is restricted to the scenario when LayerNormalization uses FusedBatchNorm and the input tensor is 2D/3D.

This fusion improves performance for inference with Transformer based language models.

### CODE REVIEW

**Comment by mdfaijul** [comment]
@qlzh727 This commit (https://github.com/tensorflow/tensorflow/commit/7172ebeac672d0bc44632edcba3504d2ddbd5ae4#diff-569c52f4e3da483a946de7b1d4a508853cd04ba1a80f653e6c709477f6be5b01) has recently removed layer-normalization from `tensorflow/python/keras/layers`. However, this PR was using this module with `from tensorflow.python.keras import layers`. I tried using `from keras import layers`, but failed. Could you please suggest me how should I change my BUILD file and dependencies?

**Comment by qlzh727** [comment]
The keras code base has been moved to keras-team/keras repository, and the code in tensorflow/python/keras will be removed soon. TF code should not rely on Keras code (since the dependency direction is wrong). Are you using keras layer for testing?

**Comment by mdfaijul** [comment]
@qlzh727 Thanks for your comment. Yes, I am using a keras layer for testing. We are trying to enable a fusion of several smaller operations generated by keras layer-normalization. This helps good performance on language models.

**Comment by qlzh727** [comment]
Thanks for the info. Since the keras code doesn't live in TF repo anymore, you might want to add the keras related test into keras repository. The grappler should have its own unit test in cpp, and the keras py test will be a e2e test.

Also cc @rohan100jain and @sun51 from tf core team. We are pushing towards using tf.function with jit_compile=True, and avoid using any specific ops/kernels from python code if possible. Your PR should be fine since it is purely graph level optimization.

File: tensorflow/core/kernels/mkl/mkl_layer_norm_op.cc [hunk 1]
```diff
@@ -0,0 +1,171 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_layer_norm_op.cc:1]: Nit:
```suggestion
/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
```

File: tensorflow/core/grappler/optimizers/remapper.cc [hunk 2]
```diff
@@ -1090,6 +1090,144 @@ bool FindMatMulBiasAddAndGelu(RemapperContext* ctx, int node_index,
...
+  // Additional check for LayerNorm
+  if (found_op_type_match) {
+    // LayerNorm uses FusedBatchNorm in training mode.
+    NodeDef* fused_batch_norm_node =
+        ctx->graph_view.GetNode(matched_nodes_map->at("fused_batch_norm"))
+            ->node();
+    bool is_training = false;
+    if (!TryGetNodeAttr(*fused_batch_norm_node, kIsTraining, &is_training))
+      return false;
+    if (!is_training) return false;
```
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper.cc:1188]: Can we merge this into
```suggestion
    if (!TryGetNodeAttr(*fused_batch_norm_node, kIsTraining, &is_training) || !is_training)
      return false;
```
? (With reformatting.)

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR! I have minor comments.

File: tensorflow/core/grappler/optimizers/remapper.cc (hunk 2 above)
Comment (mdfaijul) [reply tensorflow/core/grappler/optimizers/remapper.cc]: Done

File: tensorflow/core/kernels/mkl/mkl_layer_norm_op.cc (hunk 1 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_layer_norm_op.cc]: Done

**Comment by mdfaijul** [comment]
@penpornk Thanks for the comments. I have addressed those. Please check.

**Comment by mdfaijul** [comment]
@penpornk Mistakenly while I was clicking on comment, the click was on `Close with comment`. Reopened it. BTW, I have addressed the review comments.

File: tensorflow/core/grappler/optimizers/remapper.cc [hunk 3]
```diff
@@ -1183,9 +1183,9 @@ bool FindMklLayerNorm(RemapperContext* ctx, int node_index,
         ctx->graph_view.GetNode(matched_nodes_map->at("fused_batch_norm"))
             ->node();
     bool is_training = false;
-    if (!TryGetNodeAttr(*fused_batch_norm_node, kIsTraining, &is_training))
+    if (!TryGetNodeAttr(*fused_batch_norm_node, kIsTraining, &is_training) ||
+      !is_training)
```
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper.cc:1187]: I think the spacing is wrong here. But I'll take care of it internally.

**Review by penpornk — APPROVED** [review]
Thank you for the changes!

**Comment by mdfaijul** [comment]
@penpornk Resolved merged conflict

**Comment by gbaned** [comment]
Seems auto-merge is not happening but the changes are merged into master now, so we can close this. Thank you for the PR.

**Comment by penpornk** [comment]
This PR has been rolled back because it broke the Windows build. Below are some of the error messages:
```c++
tensorflow/core/kernels/mkl/mkl_layer_norm_op.cc(95): error C2398: Element '1': conversion from 'size_t' to 'ptrdiff_t' requires a narrowing 
tensorflow/core/kernels/mkl/mkl_layer_norm_op.cc(99): error C2398: Element '1': conversion from 'size_t' to 'ptrdiff_t' requires a narrowing conversion

tensorflow/core/kernels/mkl/mkl_layer_norm_op.cc(113): error C2036: 'void *': unknown size
```
Please feel free to open a new PR with a fix to roll this forward.
